File: Cpp/Maximum_Number_of_Distinct_Elements_After_Operations.cpp
class Solution {
public:
    int maxDistinctElements(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end());
        int cnt = 0;
        int last = INT_MIN;
        for (int n : nums) {
            int min_allowed = n - k;
            int max_allowed = n + k;
            int assign = max(min_allowed, last + 1);
            if (assign <= max_allowed) {
                last = assign;
                cnt++;
            }
        }
        return cnt;
    }
};

File: Rust/Maximum_Number_of_Distinct_Elements_After_Operations.rs
impl Solution {
    pub fn max_distinct_elements(mut nums: Vec<i32>, k: i32) -> i32 {
        nums.sort_unstable();
        let mut cnt = 0;
        let mut last = i64::MIN;
        let k64 = k as i64;
        for &n in nums.iter() {
            let num = n as i64;
            let min_allowed = num - k64;
            let max_allowed = num + k64;
            let assign = std::cmp::max(min_allowed, last + 1);
            if assign <= max_allowed {
                last = assign;
                cnt += 1;
            }
        }
        cnt
    }
}
